crates/cranelift/src/wasmfx/optimized.rs

```diff
@@ -1702,16 +1886,35 @@ pub(crate) fn translate_resume<'a>(
         );
         resume_csi.write_limits_to_vmcontext(env, builder, vm_runtime_limits_ptr);
 
-        let fiber_stack = co.get_fiber_stack(env, builder);
-        let control_context_ptr = fiber_stack.load_control_context(env, builder);
+        // Install handlers in (soon to be) parent's HandlerList:
+        // Let the i-th handler clause be (on $tag $block).
+        // Then the i-th entry of the HandlerList will be the address of $tag.
+        let handler_list = parent_csi.get_handler_list();
+
+        if resumetable.len() > 0 {
+            // If the existing list is too small, reallocate (in runtime).
+            let resumetable_len = builder.ins().iconst(I32, resumetable.len() as i64);
+            handler_list.ensure_capacity(env, builder, resumetable_len);
+
+            // Note that we currently don't remove duplicate tags.
+            let tag_addresses: Vec<ir::Value> = resumetable
+                .iter()
+                .map(|(tag_index, _)| shared::tag_address(env, builder, *tag_index))
+                .collect();
+            handler_list.store_data_entries(env, builder, &tag_addresses, false);
+        }
 
-        let resume_payload: u64 = wasmtime_continuations::ControlEffect::resume().into();
-        let resume_payload = builder.ins().iconst(I64, resume_payload as i64);
+        let resume_payload = ControlEffect::make_resume(env, builder).to_u64();
+
+        // Note that the control context we use for switching is not the one in
+        // (the stack of) resume_contref, but in (the stack of) last_ancestor!
+        let fiber_stack = last_ancestor.get_fiber_stack(env, builder);
```

Comment:
~~I don't understand the point of this "last ancestor" business. Please explain.~~ See my other comment below.